Cloud Event output resource

afrittoli opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

Implementation for a cloud event output resource: https://github.com/tektoncd/pipeline/issues/587
as specified in the design document:
https://docs.google.com/document/d/1Ry1B37QZlN6wwjgYj4TUdNwq_Wtex0A3nj11yizfqZI/edit#

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md)
for more details._

# Release Notes

<!--
Does your PR contain User facing changes?

If so, briefly describe them here so we can include this description in the
release notes for the next release!
-->

```
New output resource CloudEventPipelineResource available. 
When used in a `Task`, it will generate a cloud event upon completion of any `TaskRun` 
that references that `Task`. The cloud event includes the entire body of the `TaskRun`.
```

afrittoli commented:
/ok-to-test

afrittoli commented:
/assign @bobcatfish

afrittoli commented:
/assign @vdemeester

akihikokuroda commented on pkg/apis/pipeline/v1alpha1/cloud_event_resource.go:
```diff
@@ -0,0 +1,101 @@
+  Type           PipelineResourceType `json:"type"`
+	TargetURI      string 							`json:"targetURI"`
+	Owner 				 TaskRun	            `json:"taskRun"`
+}
```
nit: indentations

afrittoli replied:
Done

akihikokuroda commented on pkg/apis/pipeline/v1alpha1/resource_types.go:
```diff
@@ -43,10 +43,13 @@ const (
 	PipelineResourceTypePullRequest PipelineResourceType = "pullRequest"
+
+	// PipelineResourceTypeCloudEvent indicates that this source is a cloud event URI
+	PipelineResourceTypeCloudEvent PipelineResourceType = "cloudevent"
```
nit: CamelCase for consistency "cloudEvent"

afrittoli replied:
Done

akihikokuroda commented on pkg/reconciler/v1alpha1/taskrun/resources/cloudevent.go:
```diff
@@ -0,0 +1,112 @@
+
+  "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+)
+
```
nit: indentations

afrittoli replied:
Done

akihikokuroda commented on pkg/reconciler/v1alpha1/taskrun/resources/cloudevent.go:
```diff
@@ -0,0 +1,112 @@
+	cloudEventClient, err := cloudEventClientFactory(sinkURI)
+	if err != nil {
+		logger.Errorf("Error creating the cloud-event client: %s", err)
+    return event, err
```
nit: indentation

akihikokuroda commented on pkg/reconciler/v1alpha1/taskrun/resources/cloudevent.go:
```diff
@@ -0,0 +1,112 @@
+			Extensions: nil,
+		}.AsV02(),
+		Data: data,
+	}
```
Where is the "Time" set?

afrittoli replied:
It is not set, all the timestamps from the TaskRun are available in the TaskRunStatus which is sent in the body, so I did not bother setting a timestamp there. What would you set, the time that the event is created or the completion time of the TaskRun?

akihikokuroda replied:
Either is fine.  The event listener may sort or filter the events based of the time.  So I believe that it must be set.

afrittoli replied:
Ack, I'll put the TaskRun completion status then.

akihikokuroda commented on pkg/reconciler/v1alpha1/taskrun/resources/cloudevent.go:
```diff
@@ -0,0 +1,112 @@
+	_, err = cloudEventClient.Send(context.TODO(), event)
+	if err != nil {
+		logger.Errorf("Error sending the cloud-event: %s", err)
+    return event, err
```
nit: indentations

afrittoli replied:
Done

akihikokuroda commented on pkg/reconciler/v1alpha1/taskrun/resources/cloudevent.go:
```diff
@@ -0,0 +1,112 @@
+
+// SendTaskRunCloudEvent sends a cloud event for a TaskRun
+func SendTaskRunCloudEvent(sinkURI string, taskRun *v1alpha1.TaskRun, logger *zap.SugaredLogger, optionalClientFactory ...ClientFactory) (cloudevents.Event, error) {
+	// We take an optional clientFactory as input so we can mock it in tests
```
nit: indentations

afrittoli replied:
Done

akihikokuroda commented on pkg/reconciler/v1alpha1/taskrun/resources/cloudevent.go:
```diff
@@ -0,0 +1,112 @@
+  } else if taskRunStatus.IsTrue() {
+    eventType = TektonTaskRunSuccessful
+  } else {
+    return event, fmt.Errorf("Unknown condition for in TaskRun.Status %s", taskRunStatus.Status)
```
I wonder if some event should be sent here.

afrittoli replied:
That `else` case is not really like to happen, but I wanted to at least log something in case someone crafted a broken TaskRun. I would not send an event though.

akihikokuroda replied:
OK

akihikokuroda commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -330,12 +371,42 @@ func (c *Reconciler) reconcile(ctx context.Context, tr *v1alpha1.TaskRun) error
+	if sendErrors > 0 {
+		c.Logger.Errorf("Failed to send %d cloud events for TaskRun %s", sendErrors, tr.Name)
+		// Return the latest send error
+		return err
```
This returns only the error from the last error delivery.  Isn't it better to return the multierror.Error?

afrittoli replied:
Nice point, I only discovered multierror after I wrote this code - so I will adapt this one now to send all errors back.

afrittoli commented:
@akihikokuroda thank you for the review!! :)

akihikokuroda commented on pkg/reconciler/v1alpha1/taskrun/resources/cloudevent.go:
```diff
@@ -0,0 +1,112 @@
+func SendTaskRunCloudEvent(sinkURI string, taskRun *v1alpha1.TaskRun, logger *zap.SugaredLogger, optionalClientFactory ...ClientFactory) (cloudevents.Event, error) {
+	// We take an optional clientFactory as input so we can mock it in tests
+	var cloudEventClientFactory ClientFactory
+	if len(optionalClientFactory) == 0 {
```
Is this "if" statement necessary?

afrittoli replied:
I removed that bit of code, now.

afrittoli commented:
I addressed the review comment I got so far, and I changed the cloudEventClientFactory to a cloudEventClient, which hopefully it will make it easier to rebase on top of the pkg/inject refactor

afrittoli commented:
/test pull-tekton-pipeline-integration-tests

afrittoli commented:
/test pull-tekton-pipeline-integration-tests

afrittoli commented:
/test pull-tekton-pipeline-integration-tests

bobcatfish commented on docs/resources.md:
```diff
@@ -618,6 +619,213 @@ the container image
 
+### Cloud Event Resource
+
+The Cloud Event Resource represents a cloud event that is sent to a target `URI`
```
would you mind linking to some docs here about what Cloud Events are in general?

afrittoli replied:
yeah, good point, thank you

bobcatfish commented on docs/resources.md:
```diff
@@ -618,6 +619,213 @@ the container image
+- dev.tekton.event.task.failed
+
+Cloud event resources are useful to notify a third party upon the completion and
+status of a `TaskRun`. In combinations with the Tekton triggers project they can
```
maybe link to https://github.com/tektoncd/triggers ?

afrittoli replied:
done

bobcatfish commented on docs/resources.md:
````diff
@@ -618,6 +619,213 @@ the container image
+      ]
+    }
+  }
+```
````
i feel like this is a bit long to embed directly in the markdown like this 🤔  what do you think about including on a subset of it, and/or maybe including the complete example in a separate file linked from here?

afrittoli replied:
good point, I'm not sure we need a full TaskRun example - as that may become stale anyways - headers and basic data structure should be enough.

bobcatfish commented on examples/taskruns/taskrun-cloud-event.yaml:
```diff
@@ -0,0 +1,111 @@
+      resourceRef:
+        name: to-message-sink
+  taskRef:
+    name: send-cloud-event-task
```
nice example!!

bobcatfish commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -107,25 +110,42 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+			// Let's keep timeouts and sidecars running as long as we're trying to
+			// send cloud events. So we stop here an return errors encountered this far.
+			return merr.ErrorOrNil()
+		}
```
it seems like this function is getting a bit long - could some of this cloudevent specific logic be moved into a separate package perhaps? (esp since i have a feeling we might considering moving this logic around in the future)

afrittoli replied:
Uhm, most of the logic is in `sendCloudEvents`, so I could move that to a separate module, but I'm not sure that would gives us much.

There is already a nice separation between the resource and sending of cloud events - since the resource is "resolved" by adding the events to be sent to `TaskRun.Status`, and the logic in here works on the status to send the events. In future we could have a notification system write the same events to be sent to `TaskRun.Status`, and the logic here would not change a bit.

What I could do perhaps is to create a function to be invoked once the TaskRun is "done" which would be what's here moved into a dedicated function, but I wonder if I should do that once the notification system design is in place instead.

bobcatfish replied:
kk sounds good! and now that i look at this again it's only a small amount of logic really :) maybe later on itll be clearer how we can break this up, and maybe @vdemeester 's resource extensibility plans will help :D

bobcatfish commented on pkg/reconciler/v1alpha1/taskrun/resources/cloudevent/cloudevent_test.go:
```diff
@@ -0,0 +1,208 @@
+    sinkURI: defaultSinkURI,
+    eventSourceURI: invalidEventSourceURI,
+    cloudEventClient: defaultCloudEventClient,
+    wantErr: true,
```
i find that it can be a lot easier to read and maintain tests if you have a completely different test for the "wantErr" case, i.e. a set of table tests for the success case in one Test function, and another for the error cases. Then you dont need any `if c.wantErr` logic in the tests themselves

afrittoli replied:
ack, I'll split the tests.

afrittoli replied:
solved in another PR

bobcatfish commented on pkg/apis/pipeline/v1alpha1/cloud_event_resource.go:
```diff
@@ -0,0 +1,101 @@
+	Owner     TaskRun              `json:"taskRun"`
+}
+
+// NewCloudEventResource creates a new GCS resource to pass to a Task
```
i think this comment is out of sync

afrittoli replied:
nice catch, thanks

afrittoli replied:
solved in a different PR

bobcatfish commented on pkg/apis/pipeline/v1alpha1/cloud_event_resource.go:
```diff
@@ -0,0 +1,101 @@
+	corev1 "k8s.io/api/core/v1"
+)
+
+// CloudEventResource is an event sink to which events are delivered when a Task has finished
```
nit: when a TaskRun has finished

afrittoli replied:
solved in a different PR

bobcatfish commented on docs/resources.md:
```diff
@@ -618,6 +619,213 @@ the container image
+  type: cloudevent
+  params:
+  - name: targetURI
+    value: http://sink:8080
```
i think we might want some kind of "nullsink" type, so that folks can use Tasks that take this PipelineResource, even if they don't actually have a sink?

(probably something for a follow up PR tho!)

afrittoli replied:
Oh, nice idea. 
Another approach could be to support - in general - optional input resources - which is something that has been mentioned before, to make a Task more re-usable.

bobcatfish replied:
ooo that's an interesting idea too!

afrittoli commented:
> Very cool @afrittoli !! I took an overall look but haven't yet been able to look into all of it in detail. I'm wondering if there is a way this could happen over the course of several PRs, to speed up the process for you and make sure each change gets reviewed carefully enough? One idea of how this could be broken up a bit:
> 
>     1. A PR which just adds the types to the spec definitions (this _would_ mean potentially a release where these exist but do nothing - personally I'm okay with it tho! + @vdemeester @dlorenc @ImJasonH @abayer )
> 
>     2. A PR which adds pkg/apis/pipeline/v1alpha1/cloud_event_resource.go and tests
> 
>     3. A PR that adds pkg/reconciler/v1alpha1/taskrun/resources/cloudevent/cloudevent.go  and tests
> 
>     4. Finally a PR that updates the reconciler to actually hook everything up + make the logic happen (which would include the docs)
> 
> 
> Some of 1-3 could probably happen in parallel, depending on what works for you!
> 
> Or feel free to break this up some other way - if at all possible I think the reviews would be much easier and go much faster is we can break it put 🙏 (If not then I will just have to get a chunk of time when I can give this my full attention for 1hr+! Will make turnarounds slower tho)

As long as folks are OK with having code in the repo which is not actually used, splitting the way you proposed should be pretty straight forward. The concern of having unused code in the repo was what stopped me from splitting the PR in multiple pieces to begin with, but as long as we recommend our users to pick-up use releases, then it should be fine.

bobcatfish commented:
> As long as folks are OK with having code in the repo which is not actually used, splitting the way you proposed should be pretty straight forward

I hear you! I have mixed feelings about it myself - especially if we got interrupted or something and never added the code that actually uses it 😩 

Personally I think it's worth it but I'll ping the other owners again to get a couple other opinions!

akihikokuroda commented on pkg/reconciler/v1alpha1/taskrun/resources/cloudevent/cloudevent.go:
```diff
@@ -0,0 +1,113 @@
+
+const (
+	// TektonTaskRunUnknown is sent for TaskRuns with "ConditionSucceeded" "Unknown"
+	TektonTaskRunUnknown TektonEventType = "dev.tekton.event.task.unknown"
```
Would you add the version to these types?

> Type: String
> Description: This attribute contains a value describing the type of event related to the originating occurrence. Often this attribute is used for routing, observability, policy enforcement, etc. The format of this is producer defined and might include information such as the version of the type - see Versioning of Attributes in the Primer for more information https://github.com/cloudevents/spec/blob/master/primer.md#versioning-of-attributes

afrittoli replied:
Yeah, the event types become part of the interface, so we'll need a way to version them.
I'll address this in the dedicated PR

afrittoli replied:
https://github.com/tektoncd/pipeline/pull/1112

akihikokuroda commented on pkg/reconciler/v1alpha1/taskrun/resources/cloudevent/cloudevent.go:
```diff
@@ -0,0 +1,113 @@
+		return event, fmt.Errorf("Unknown condition for in TaskRun.Status %s", taskRunStatus.Status)
+	}
+	eventSourceURI := taskRun.ObjectMeta.SelfLink
+	data, _ := json.Marshal(taskRun)
```
Would you make the taskrun data tagged as like "taskrun" instead of the taskrun data as the whole data.  I think that it makes the future extension easier.  I'm thinking to add the correlator or some arbitrary information from the cloudevent resource like the pull request URL.

afrittoli replied:
uhm, that's a good idea, thank you!

afrittoli replied:
https://github.com/tektoncd/pipeline/pull/1112

afrittoli commented:
> > As long as folks are OK with having code in the repo which is not actually used, splitting the way you proposed should be pretty straight forward
> 
> I hear you! I have mixed feelings about it myself - especially if we got interrupted or something and never added the code that actually uses it 😩
> 
> Personally I think it's worth it but I'll ping the other owners again to get a couple other opinions!

I started preparing four different PRs as you suggested. The first three are independent and can run in parallel (or at least I think so, let's see what CI says), and the last one needs to put all together.
How do I express this dependency reasonably with GitHub though?
I cherry-pick the first three commits and put one more on top and push all together. Gerrit would recognize the initial commits as patches it knows about, but as far as I've seen GitHub will just treat it as a separate PR. Is there a best practice for how to handle stacked patches in GitHub?

afrittoli commented:
This is a first attempt to split this PR. 
The first three commits are in dedicated PRs: https://github.com/tektoncd/pipeline/pull/1090, https://github.com/tektoncd/pipeline/pull/1091 and https://github.com/tektoncd/pipeline/pull/1092.
I will address review comments for the top commit here and there rest in the specific PRs.

Once the other PRs are approved this needs to be rebased and restested, so I will mark it as on hold for now.

afrittoli commented:
/hold

afrittoli commented:
/hold wip

afrittoli commented:
/hold cancel

bobcatfish commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -108,25 +111,42 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
 	}
 
 	if tr.IsDone() {
+		var merr *multierror.Error
```
hmm interesting!

bobcatfish replied:
might be worth a comment explaining why we are using multierror in this case, when we don't elsewhere? (reading through im thinking its b/c handling cloudevents is kind of "in addition" to our "main" logic, so we want to track errors related to that but still continue execution?)

afrittoli replied:
Added a comment

bobcatfish commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -108,25 +111,42 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+		merr = multierror.Append(cloudEventErr, updateErr)
+		if cloudEventErr != nil {
+			// Let's keep timeouts and sidecars running as long as we're trying to
+			// send cloud events. So we stop here an return errors encountered this far.
```
nice comments!

bobcatfish commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -293,12 +331,41 @@ func (c *Reconciler) reconcile(ctx context.Context, tr *v1alpha1.TaskRun) error
 	return nil
 }
 
+func (c *Reconciler) sendCloudEvents(tr *v1alpha1.TaskRun) error {
```
i have a feeling im repeating myself in which case i apologize but i really would like to avoid adding functions to the reconciler as much as possible - id rather see this as a function in a separate package called by the reconcile function. it seems like the reconciler is providing only a logger and a client which should be pretty easy to pass in instead of giving the function full access to the reconciler

afrittoli replied:
done

bobcatfish commented on pkg/reconciler/v1alpha1/taskrun/taskrun_test.go:
```diff
@@ -1616,3 +1636,195 @@ func TestHandlePodCreationError(t *testing.T) {
 	}
 }
+
+func TestReconcileCloudEvens(t *testing.T) {
```
typo: Events

afrittoli replied:
done

afrittoli commented:
Fixed a small bug, moved some logic out of the taskrun controller to a dedicated module, added some test coverage, rebased. 
There is space for some more test coverage, I can do it on top or separate PR.

afrittoli commented:
/retest

afrittoli commented:
Added extra unit test coverage, fixed integration test.

bobcatfish commented on pkg/reconciler/v1alpha1/taskrun/resources/cloudevent/cloud_event_controller.go:
```diff
@@ -0,0 +1,97 @@
+}
+
+func cloudEventDeliveryFromTargets(targets []string) []v1alpha1.CloudEventDelivery {
+	// len(nil slice) is 0
```
hm im not sure this comment is needed?

afrittoli replied:
nice catch, it is definitely not, leftover...

bobcatfish commented on pkg/reconciler/v1alpha1/taskrun/resources/cloudevent/cloud_event_controller.go:
```diff
@@ -0,0 +1,97 @@
+// InitializeCloudEvents initializes the CloudEvents part of the
+// TaskRunStatus from a slice of PipelineResources
+func InitializeCloudEvents(tr *v1alpha1.TaskRun, prs []*v1alpha1.PipelineResource) {
+	// FIXME(afrittoli) If there are no events this is run every time
```
is there a github issue we could put in this FIXME as well? I'm also not quite sure what the implication is - does it mean the code is hit every time (which seems okay?) or does something actually happen that we don't want to have happen?

afrittoli replied:
In case no cloud events are defined, the loop in L34-39 will be executed on every reconcile - not a big deal really :)

bobcatfish commented on pkg/reconciler/v1alpha1/taskrun/resources/cloudevent/cloud_event_controller_test.go:
```diff
@@ -0,0 +1,251 @@
+	}
+}
+
+// TBD: TestSendCloudEvents error tests
```
would be good to put an issue # here too, like `TODO(#111111)` (otherwise stuff like this can just kinda accumulate and itll not be clear later on what action should be taken)

afrittoli replied:
I forgot to remove this comment, it's done in `TestSendCloudEventsErrors`

bobcatfish commented:
/lgtm
/meow space
